jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java

```diff
@@ -170,7 +170,11 @@ public static boolean isSameName(Path pathA, Path pathB)
     @Override
     public boolean exists()
     {
-        return Files.exists(targetPath != null ? targetPath : path);
+        if (Files.exists(path))
+            return true;
+        if (isAlias())
+            return Files.exists(targetPath);
+        return false;
```

Comment:
`Files.exists()` implicitly take a `FOLLOW_SYMLINKS`, so the first at line 173 check should be enough.

Reply:
Not if the path has unresolved `..` segments, then `Files.exists()` returns false even if it does exist.

Reply:
I agree with sbordet

If `this.path` was created with one of the `URI` based methods for creating a `Path` (eg: `Paths.of(URI)` or `Paths.get(URI)`) then your statement is true.
As there is an expectation that the `URI` it is given is absolute, canonical, and normalized.
But if `this.path` was created from a `String` or other `Path` object, even the `/../` segments have no impact on the `Files.exists()` logic and will return true.

Reply:
> But if this.path was created from a String or other Path object, even the /../ segments have no impact on the Files.exists() logic and will return true.

This is not correct.

The following test passes and shows that Files.exists does not work with `/../` segments. And since you can create `PathResource` with a `Path` directly we cannot assume that the path does not contain `/../` segments.
```java
Path resourcePath = testDir.resolve("resource.txt");
assertTrue(Files.exists(resourcePath));

Path aliasPath = testDir.resolve("foo/../resource.txt");
assertFalse(Files.exists(aliasPath));
assertTrue(Files.exists(aliasPath.normalize()));
```

Reply:
The `Files.exists()` does work with `/../` if each element of the path exists.
In your case `foo` does not exist on your filesystem.

This is part of the filesystem specific normalization techniques.
Some filesystems will find it (some will not).
Don't write your code assuming that all filesystems behave the same in this regard.

Example code.

``` java
package fs;

import java.io.IOException;
import java.net.URI;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

import org.eclipse.jetty.toolchain.test.FS;

public class PathsGetExample
{
    public static void main(String[] args) throws IOException
    {
        boolean withFoo = true;
        System.out.printf("withFoo = %b%n", withFoo);

        // Create some content
        Path testDir = Paths.get("/tmp/paths-get-examples");
        FS.ensureDirExists(testDir);
        Path fooDir = testDir.resolve("foo");
        if (withFoo)
            FS.ensureDirExists(testDir.resolve("foo"));
        else
            Files.deleteIfExists(fooDir);
        FS.touch(testDir.resolve("test.txt"));

        URI uri = URI.create(testDir.toUri().toASCIIString() + "foo/../test.txt");
        Path a = Paths.get(uri);
        System.out.printf("%s -> %s -> %b%n", uri, a, Files.exists(a));

        URI urin = uri.normalize();
        Path b = Paths.get(urin);
        System.out.printf("%s -> %s -> %b%n", urin, b, Files.exists(b));

        String str = "/tmp/paths-get-examples/foo/../test.txt";
        Path c = Paths.get(str);
        System.out.printf("%s -> %s -> %b%n", str, c, Files.exists(c));
    }
}
```

Executing with "foo"

```
withFoo = true
file:///tmp/paths-get-examples/foo/../test.txt -> /tmp/paths-get-examples/foo/../test.txt -> true
file:/tmp/paths-get-examples/test.txt -> /tmp/paths-get-examples/test.txt -> true
/tmp/paths-get-examples/foo/../test.txt -> /tmp/paths-get-examples/foo/../test.txt -> true
```

Executing without "foo"

```
withFoo = false
file:///tmp/paths-get-examples/foo/../test.txt -> /tmp/paths-get-examples/foo/../test.txt -> false
file:/tmp/paths-get-examples/test.txt -> /tmp/paths-get-examples/test.txt -> true
/tmp/paths-get-examples/foo/../test.txt -> /tmp/paths-get-examples/foo/../test.txt -> false
```

Reply:
This is the same behavior you get on linux with `ls` ...

```
[joakim@hyperion ~]$ ls -la /tmp/paths-get-examples/
total 512
drwxrwxr-x   2 joakim joakim   4096 Oct 14 07:45 ./
drwxrwxrwt 221 root   root   516096 Oct 14 07:42 ../
-rw-rw-r--   1 joakim joakim      0 Oct 14 07:42 test.txt

[joakim@hyperion ~]$ ls -la /tmp/paths-get-examples/foo/../test.txt
ls: cannot access '/tmp/paths-get-examples/foo/../test.txt': No such file or directory

[joakim@hyperion ~]$ mkdir /tmp/paths-get-examples/foo

[joakim@hyperion ~]$ ls -la /tmp/paths-get-examples/
total 516
drwxrwxr-x   3 joakim joakim   4096 Oct 14 07:45 ./
drwxrwxrwt 221 root   root   516096 Oct 14 07:42 ../
drwxrwxr-x   2 joakim joakim   4096 Oct 14 07:45 foo/
-rw-rw-r--   1 joakim joakim      0 Oct 14 07:42 test.txt

[joakim@hyperion ~]$ ls -la /tmp/paths-get-examples/foo/../test.txt
-rw-rw-r-- 1 joakim joakim 0 Oct 14 07:42 /tmp/paths-get-examples/foo/../test.txt
```

Reply:
The way I see it we have a few options here.

1. We have all `Resource` implementations do the URI normalization, removing segments like `/../` before creating the `Path` object.
2. We have all `Resource` implementations honor the as-submitted `URI`, with no `Resource` implementation normalization, and let the FileSystem suss out if the content can be referenced or not.

We are doing 1 currently, but should we?
Is it such a terrible idea to reject requests for content like `dir-does-not-exist/../test.dat` and allow `dir-that-exists/../test.dat`?
I could easily be convinced that the `Resource` implementations should not perform URI normalization.
It just means a few types of torturous requests we used to receive in the past we no longer support.

Or, if we want to support these requests still, it means that the normalization occurs at the layer above `Resource.resolve(String)` (which kinda makes sense, and I suspect it would simplify the protected paths issues too)